refbox/gamemanager.py

```diff
@@ -111,3 +124,80 @@ def timeoutStateRef(self):
 
     def setTimeoutStateRef(self):
         self._timeout_state = TimeoutState.ref
+
+    def addPenalty(self, p):
+        self._penalties[p.team()].append(p)
+
+    def delPenalty(self, p):
+        if p in self._penalties[p.team()]:
+            self._penalties[p.team()].remove(p)
+
+    def penalties(self, team_color):
+        return self._penalties[team_color]
+
+    def deleteAllPenalties(self):
+        self._penalties = [[],[]]
+
+    def _start_unstarted_penalties(self, game_clock):
+        for p in self._penalties[TeamColor.white] + self._penalties[TeamColor.black]:
+            if not p.startTime():
+                p.setStartTime(game_clock)
+
+    def _resetOutstandingPenalties(self):
+        """ Reset any penalties that were interrupted by half-time.
+        """
+        for p in self._penalties[TeamColor.white] + self._penalties[TeamColor.black]:
+            if not p.servedCompletely(0, GameState.first_half):
+                p._duration_remaining = p.timeRemaining(0, GameState.first_half)
+                p._start_time = self._duration
+
+
+class Penalty(object):
+
+    def __init__(self, player, team, duration, start_time = None):
+        self._player = player
+        self._team = team
+
+        # Game time when the penalty started
+        self._start_time = start_time
+
+        # Total time of the penalty
+        self._duration = duration
```

Comment:
Why keep this around?

Reply:
It's an artifact of a part of the UI flow that I haven't figured out how to handle well yet:

Timeout -> add penalty -> submit -> resume -> half time -> second half -> edit penalty

The edit window needs to know how long the penalty was supposed to be, so it can pre-select the relevant duration. However, the player has already served some amount of the time by the time we got there.